ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -293,6 +316,186 @@ public TransientQueryMetadata executeTransientQuery(
+    final ConfiguredStatement<Query> statement = statementOrig.withConfigOverrides(
+        ImmutableMap.<String, Object>builder()
+            .putAll(statementOrig.getSessionConfig().getOverrides())
+            .put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+            .put(StreamsConfig.NUM_STREAM_THREADS_CONFIG, 1)
+            .put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100)
+            .build()
+    );
+    try {
+
+      final TransientQueryMetadata transientQueryMetadata = EngineExecutor
+          .create(primaryContext, serviceContext, statement.getSessionConfig())
+          .executeTransientQuery(statement, excludeTombstones);
+
+      QueryLogger.info(
+          "Streaming stream pull query results '{}'",
+          statement.getStatementText()
+      );
+
+      return transientQueryMetadata;
+
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      // add the statement text to the KsqlException
+      throw new KsqlStatementException(e.getMessage(), statement.getStatementText(), e.getCause());
+    }
+  }
+
+  public void waitForStreamPullQuery(
+      final ServiceContext serviceContext,
+      final ImmutableAnalysis analysis,
+      final ConfiguredStatement<Query> statement,
+      final TransientQueryMetadata transientQueryMetadata) {
+    // query the endOffsets of the input
+    final String sourceTopicName = analysis.getFrom().getDataSource().getKafkaTopicName();
+    final Admin admin = serviceContext.getAdminClient();
+    final TopicDescription topicDescription = getTopicDescription(
+        admin,
+        sourceTopicName
+    );
+
+    final Object processingGuarantee = statement
+        .getSessionConfig()
+        .getConfig(true)
+        .getKsqlStreamConfigProps()
+        .get(StreamsConfig.PROCESSING_GUARANTEE_CONFIG);
+
+    final IsolationLevel isolationLevel =
+        StreamsConfig.AT_LEAST_ONCE.equals(processingGuarantee)
+            ? IsolationLevel.READ_UNCOMMITTED
+            : IsolationLevel.READ_COMMITTED;
+
+    final Map<TopicPartition, Long> endOffsets = getEndOffsets(
+        admin,
+        topicDescription,
+        isolationLevel
+    );
+
+    // wait for the query to pass the endOffsets
+    while (!passedEndOffsets(admin, transientQueryMetadata, endOffsets)) {
+      if (transientQueryMetadata.getState() == State.ERROR) {
+        log.error("Stream pull query execution has failed");
+        throw new KsqlServerException("Server error");
+      } else {
+        try {
+          Thread.sleep(50L);
+        } catch (final InterruptedException e) {
+          log.error("Interrupted waiting for stream pull query to complete", e);
+          throw new KsqlServerException("Interrupted");
+        }
+      }
+    }
+    transientQueryMetadata.getKafkaStreams().close(Duration.ZERO);
+    QueryLogger.info("Finished pull query results", statement.getStatementText());
```

Comment:
Yep, it gets registered in the StreamedQueryResource when we do this:
```
      // stop the response stream before sending the successful response.
      queryStreamWriter.close();
```

Reply:
Now that I'm thinking about it again, IIRC, the proactive close here is actually just to make the query termination faster. Otherwise, I think it had to wait for some component's polling interval to close the query asynchronously. But since we're waiting on the query synchronously in this case, it's beneficial to start the process of closing KS immediately. Admittedly a bit hacky, but then again this whole implementation strategy is more or less a hack.